The HTU21 sensor bring up

vadimol opened the pull request:
1) Added sensor directorty to dev
2) Added htu21 sensor implementation

vadimol committed:
1) Added sensor directorty to dev
2) Added htu21 sensor implementation

forGGe commented on dev/sensor/htu21/export/dev/sensor/htu21.hpp:
```diff
@@ -0,0 +1,181 @@
+#ifndef __HTU21_HPP__
```
Let's reflect component path in define name. I.e.:
__DEV_SENSOR_HTU21_HPP__

forGGe replied:
Add file header in doxygen style

forGGe commented on dev/sensor/htu21/export/dev/sensor/htu21.hpp:
```diff
@@ -0,0 +1,181 @@
+
+namespace sensor {
+
+template <class i2c_dev>
```
Add brief description of a class in doxygen style

forGGe commented on dev/sensor/htu21/export/dev/sensor/htu21.hpp:
```diff
@@ -0,0 +1,181 @@
+    //! \brief Reads sample from sensor and converts to a physical value.
+    //! \retval Temperature in C degree.
+    //!
+    static float get_temperature();
```
That's dangerous. Not all hardaware have hardware floating point unit, and not all HW works fast. Let's use an integer, interpreted as real number. Say, value of 12345 will mean 12.345 degree in Celsius.

forGGe commented on dev/sensor/htu21/export/dev/sensor/htu21.hpp:
```diff
@@ -0,0 +1,181 @@
+    static float get_humidity();
+
+    //! Sensor I2C address. Cannot be changed.
+    static constexpr uint8_t i2_addr = 0x80;
```
Wondering if it worth keep it public

forGGe commented on dev/sensor/htu21/export/dev/sensor/htu21.hpp:
```diff
@@ -0,0 +1,181 @@
+    //! \brief Writes to the sensor user register.
+    //! \details Allows to configure sensor resolution.
+    //!
+    static void write_user_register(uint8_t value);
```
Let's discuss a bit more this.

vadimol committed:
Merge branch 'develop' of https://github.com/forGGe/theCore into htu21_sensor

vadimol committed:
1) Addressed review comments
2) Done code clean up
3) Added advanced sensor methods
4) Added additional doxygen

vadimol commented on dev/sensor/htu21d/export/dev/sensor/htu21d.hpp:
```diff
@@ -0,0 +1,327 @@
+}
+
+template <class i2c_dev>
+typename htu21d<i2c_dev>::resolution_t htu21d<i2c_dev>::get_resolution_mode()
```
Not quite sure if it is fine, Max, what do you think?

vadimol committed:
Merge branch 'develop' into htu21_sensor

vadimol committed:
1) added enum class for sensor resolution
2) added error handling

vadimol commented:
addressed review comments

forGGe commented on dev/sensor/htu21d/export/dev/sensor/htu21d.hpp:
```diff
@@ -0,0 +1,427 @@
+    // since they contain status information
+    sample &= ~3;
+
+    value = -46850 + ((175000 * static_cast<long long>(sample)) / (1 << 16));
```
Use uint64_t for better readability.

vadimol replied:
done

forGGe commented on dev/sensor/htu21d/export/dev/sensor/htu21d.hpp:
```diff
@@ -0,0 +1,427 @@
+#ifndef __DEV_SENSOR_HTU21D_HPP__
+#define __DEV_SENSOR_HTU21D_HPP__
+
+namespace sensor {
```
Add another namespace above 'sensor': 'ecl'

vadimol replied:
done

forGGe commented on dev/sensor/htu21d/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+add_library(htu21d INTERFACE)
```
Lacks dependency on 'types' library

vadimol replied:
could you put more details regarding this, please

forGGe replied:
Specifically, you must specify that this library depends on `types` library which contains `enum class err`:
`target_link_libraries(htu21d INTERFACE types)

forGGe commented on dev/sensor/htu21d/export/dev/sensor/htu21d.hpp:
```diff
@@ -0,0 +1,427 @@
+//! \brief Driver for HTU21D digital humidity sensor with temperature output
+//!  See, http://www.meas-spec.com/downloads/HTU21D.pdf
+//!
+
```
`#include <ecl/err.hpp>`

vadimol replied:
done

forGGe commented:
Few comments, looks good otherwise.

vadimol committed:
1) Wrapped sensor namespace into ecl
2) Added types library deps

forGGe commented:
Looks good
